hax/hax/hax.py

```diff
@@ -122,14 +122,20 @@ def main():
         stats_updater = _run_stats_updater_thread(motr, consul_util=util)
         # [KN] This is a blocking call. It will work until the program is
         # terminated by signal
+        # stats_updater and rconfc_started both invoke
+        # {adopt,shun}_motr_thread(), adopt_motr_thread() takes a reference
+        # on motr addb2 machine internally in motr and shun_motr_thread()
+        # releases it. We are invoke motr::stop() and motr::fini() when
+        # consumer thread stops, thus, it is important that all the motr
+        # references are dropped before invoking motr finalization.
+        # Hence, consumer thread is last to stop and join.
```

Comment:
I'm afraid the order is still not guaranteed. Firstly `stop()` method is invoked for every thread, then we iterate again and perform `join()`. `stop()` is not blocking, it actually can't guarantee that the thread will stop immediately. That's why relying on the order of threads in the `threads_to_wait` list is not redundant. If we do need to guarantee the order between the threads, we'll most probably need to use other mechanisms (such as barriers or something).

Reply:
You are right, after sending stop signal each thread can finalize out of order. We could move motr_thread_shut() from filestats_updater and rconfc_starter to `stop_iteration` section but in case of exception it might be possible that `motr_thread_shun` is not invoked. Let me see and find a better place to do so.

Reply:
Moved adopt and shun to hax.c to their respective functions. Restored the order of the threads.